src/mongocxx/collection.hpp

```diff
@@ -608,34 +609,9 @@ template <typename document_view_iterator_type>
 MONGOCXX_INLINE stdx::optional<result::insert_many> collection::insert_many(
     document_view_iterator_type begin, document_view_iterator_type end,
     const options::insert& options) {
-    class bulk_write writes(options.ordered().value_or(true));
-
-    result::insert_many::id_map inserted_ids{};
-    std::size_t index = 0;
-    std::for_each(begin, end, [&](const bsoncxx::document::view& current){
-        // TODO: put this somewhere else not in header scope (bsoncxx::builder)
-        if ( !current["_id"]) {
-            bsoncxx::builder::stream::document new_document;
-            new_document << "_id" << bsoncxx::oid(bsoncxx::oid::init_tag)
-                         << bsoncxx::builder::stream::concatenate{current};
-
-            writes.append(model::insert_one(new_document.view()));
-
-            inserted_ids.emplace(index++, new_document.view()["_id"]);
-        } else {
-            writes.append(model::insert_one(current));
-
-            inserted_ids.emplace(index++, current["_id"]);
-        }
-
-    });
+    insert_many_builder op = std::for_each(begin, end, insert_many_builder{options});
```

Comment:
This is... subtle. So insert_many_builder has operator(), and accumulates state, and is then moved out of the std::for_each?.

That is kinda amazing. I would not have expected that to work.

My suggestion though would be to capture op by auto:

auto op = ...